src/profile-logic/sanitize.js

```diff
@@ -362,6 +374,125 @@ function sanitizeThreadPII(
+    // sanitized, with their functions as a key, so that we can easily change
+    // all frames if we need to.
+    const sanitizedFuncIndexesToFrameIndex: Map<
+      IndexIntoFuncTable,
+      IndexIntoFrameTable[]
+    > = new Map();
+    // This set holds all func indexes that shouldn't be sanitized. This will be
+    // intersected with the previous map's keys to know which functions need to
+    // be split in 2.
+    const funcIndexesToBeKept = new Set();
+
+    const { frameTable, funcTable, resourceTable } = newThread;
+    for (let frameIndex = 0; frameIndex < frameTable.length; frameIndex++) {
+      const innerWindowId = frameTable.innerWindowID[frameIndex];
+      const funcIndex = frameTable.func[frameIndex];
+      if (innerWindowId !== null && windowIdToBeSanitized.has(innerWindowId)) {
+        // The function pointed by this frame should be sanitized.
+        let sanitizedFrameIndexes =
+          sanitizedFuncIndexesToFrameIndex.get(funcIndex);
+        if (!sanitizedFrameIndexes) {
+          sanitizedFrameIndexes = [];
+          sanitizedFuncIndexesToFrameIndex.set(
+            funcIndex,
+            sanitizedFrameIndexes
+          );
+        }
+        sanitizedFrameIndexes.push(frameIndex);
+      } else {
+        // The function pointed by this frame whould be kept.
+        funcIndexesToBeKept.add(funcIndex);
+      }
+    }
+
+    const resourcesToBeSanitized = new Set();
+
+    if (sanitizedFuncIndexesToFrameIndex.size) {
+      const newFuncTable = (newThread.funcTable =
+        shallowCloneFuncTable(funcTable));
+      const newFrameTable = (newThread.frameTable = {
+        ...frameTable,
+        func: frameTable.func.slice(),
+        line: frameTable.line.slice(),
+        column: frameTable.column.slice(),
+      });
+
+      for (const [
+        funcIndex,
+        frameIndexes,
+      ] of sanitizedFuncIndexesToFrameIndex.entries()) {
+        if (funcIndexesToBeKept.has(funcIndex)) {
+          // This function is used by both private and non-private data, therefore
+          // we split this function into 2 sanitized and unsanitized functions.
+          const sanitizedFuncIndex = newFuncTable.length;
+          newFuncTable.name.push(stringArray.length);
+          stringArray.push(`<Func #${sanitizedFuncIndex}>`);
+          newFuncTable.isJS.push(funcTable.isJS[funcIndex]);
+          newFuncTable.relevantForJS.push(funcTable.isJS[funcIndex]);
+          newFuncTable.resource.push(-1);
+          newFuncTable.fileName.push(null);
+          newFuncTable.lineNumber.push(null);
+          newFuncTable.columnNumber.push(null);
+          newFuncTable.length++;
+
+          frameIndexes.forEach(
+            (frameIndex) =>
+              (newFrameTable.func[frameIndex] = sanitizedFuncIndex)
+          );
+        } else {
+          // This function is used only by private data, so we can change it
+          // directly.
+          newFuncTable.name[funcIndex] = stringArray.length;
```

Comment:
Should we remove the old string array value as well? But I think it's a bit tricky too because we need to figure out if it's used somewhere else as well..

Reply:
Yeah, I thought this was a lot of work. Also I thought that just a function name isn't so private to keep, especially when it's burried deep inside the string table.

Ideally we should probably rebuild the profile completely, by iterating every thread and every object, and recreate the string table from scratch... Maybe we'll have to do that eventually.